### PR TITLE
[FSTORE-1331] Do not disable Hopsworks Feature Query Service on retryable failure

### PR DESCRIPTION
**Pull request by vatj**
…ient

This PR adds/fixes/changes...
- please summarize your changes to the code 
- and make sure to include all changes to user-facing APIs

JIRA Issue: -

Priority for Review: -

Related PRs: -

**How Has This Been Tested?**

- [ ] Unit Tests
- [ ] Integration Tests
- [ ] Manual Tests on VM

**Checklist For The Assigned Reviewer:**

```
- [ ] Checked if merge conflicts with master exist
- [ ] Checked if stylechecks for Java and Python pass
- [ ] Checked if all docstrings were added and/or updated appropriately
- [ ] Ran spellcheck on docstring
- [ ] Checked if guides & concepts need to be updated
- [ ] Checked if naming conventions for parameters and variables were followed
- [ ] Checked if private methods are properly declared and used
- [ ] Checked if hard-to-understand areas of code are commented
- [ ] Checked if tests are effective
- [ ] Built and deployed changes on dev VM and tested manually
- [x] (Checked if all type annotations were added and/or updated appropriately)
```

### CODE REVIEW

File: python/hsfs/core/arrow_flight_client.py [hunk 1]
```diff
@@ -46,84 +52,201 @@ def close():
...
-            location=host_url,
+            location=self.host_url,
             tls_root_certs=tls_root_certs,
             cert_chain=cert_chain,
             private_key=private_key,
             override_hostname="flyingduck.service.consul",
+            generic_options=[
+                (
+                    # https://arrow.apache.org/docs/cpp/flight.html#excessive-traffic
+                    "GRPC_ARG_MIN_RECONNECT_BACKOFF_MS",
```
Comment (kennethmhc) [inline python/hsfs/core/arrow_flight_client.py:205]: how many times will it retry?

File: python/hsfs/core/arrow_flight_client.py [hunk 2]
```diff
@@ -191,52 +328,81 @@ def afs_error_handler_wrapper(instance, *args, **kw):
...
                         and "Please register client certificates first." in message
                     ):
                         instance._register_certificates()
                         return func(instance, *args, **kw)
+                    elif (
+                        isinstance(e, FlightServerError)
+                        and "no free slot available for" in message
+                    ):
+                        raise FeatureStoreException(
+                            "No free slots available in Hopsworks Feature Query Service. Please try again later."
```
Comment (kennethmhc) [inline python/hsfs/core/arrow_flight_client.py:344]: imo `Hopsworks Feature Query Service is busy right now` is easier to understand.

File: python/hsfs/core/arrow_flight_client.py [hunk 3]
```diff
@@ -397,3 +567,55 @@ def supports(self, featuregroups):
...
+
+    @property
+    def missing_variable_name(self) -> Optional[str]:
+        """Name of the missing variable that caused the client to be disabled, if any."""
+        return self._missing_variable_name
+
+    @property
+    def health_check_succeeded_once(self) -> bool:
+        """Whether the healthcheck has succeeded at least once."""
+        return self._health_check_succeeded_once
```
Comment (kennethmhc) [inline python/hsfs/core/arrow_flight_client.py:616]: are these two properties necessary?

File: python/hsfs/core/arrow_flight_client.py [hunk 4]
```diff
@@ -46,84 +52,201 @@ def close():
...
-                )
+                self._missing_variable_name = "service_discovery_domain"
+                return None
             host_url = f"grpc+tls://flyingduck.service.{service_discovery_domain}:5005"
+        _logger.debug(
+            f"Connecting to Hopsworks Feature Query Service on host {host_url}"
+        )
+        return host_url
 
+    def _disable_for_session(self, message: Optional[str] = None) -> None:
```
Comment (kennethmhc) [inline python/hsfs/core/arrow_flight_client.py:175]: not used right now

File: python/hsfs/core/arrow_flight_client.py (hunk 1 above)
Comment (vatj) [reply python/hsfs/core/arrow_flight_client.py:205]: This does not trigger retries, however it limits the possibility of retrying faster than that

File: python/hsfs/core/arrow_flight_client.py (hunk 4 above)
Comment (vatj) [reply python/hsfs/core/arrow_flight_client.py:175]: ?

File: python/hsfs/core/arrow_flight_client.py (hunk 3 above)
Comment (vatj) [reply python/hsfs/core/arrow_flight_client.py:616]: It makes for better error messages. One is solvable only via admin intervention, the other can be temporary client or cluster issues independent of cluster config. So I would say, yes. But ready and healthcheck_succeeded_once are maybe not as valuable

File: python/hsfs/core/arrow_flight_client.py (hunk 2 above)
Comment (vatj) [reply python/hsfs/core/arrow_flight_client.py:344]: Ok

File: python/hsfs/core/arrow_flight_client.py (hunk 3 above)
Comment (vatj) [reply python/hsfs/core/arrow_flight_client.py:616]: Ok no I see your point, I can merge them in one. I will do that. Thanks for the feedback

File: python/hsfs/core/arrow_flight_client.py (hunk 4 above)
Comment (kennethmhc) [reply python/hsfs/core/arrow_flight_client.py:175]: overlooked

File: python/hsfs/core/arrow_flight_client.py (hunk 1 above)
Comment (vatj) [reply python/hsfs/core/arrow_flight_client.py:196]: This does not retry, it just prevents retry to be too earger

File: python/hsfs/core/arrow_flight_client.py [hunk 5]
```diff
@@ -46,130 +54,253 @@ def close():
...
+            self._health_check()
+            self._register_certificates()
+        except Exception as e:
+            _logger.debug("Failed to connect to Hopsworks Feature Query Service")
+            _logger.exception(e)
+            warnings.warn(
+                f"Failed to connect to Hopsworks Feature Query Service, got {str(e)}."
+                + self.DEFAULTING_TO_DIFFERENT_SERVICE_WARNING
+                + self.CLIENT_WILL_STAY_ACTIVE_WARNING
+            )
```
Comment (kennethmhc) [inline python/hsfs/core/arrow_flight_client.py:189]: need to call `_disable_for_session` here?

**Review by kennethmhc — COMMENTED** [review]
Look good!

File: python/hsfs/core/arrow_flight_client.py (hunk 5 above)
Comment (vatj) [reply python/hsfs/core/arrow_flight_client.py:202]: No it is what the message is saying, the client will stay active and they can explicitly override the behaviour.

**Review by kennethmhc — APPROVED** [review]
LGTM!

**Pull request merged**